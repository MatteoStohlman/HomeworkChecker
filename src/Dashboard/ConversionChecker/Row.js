import React from 'react';
import {compose,withState,withHandlers} from 'recompose';
import { withStyles } from '@material-ui/core/styles';
import TableCell from '@material-ui/core/TableCell';
import TableRow from '@material-ui/core/TableRow';
import UnitPicker from 'Components/UnitPicker'
import TextField from '@material-ui/core/TextField';
import math from 'mathjs'
import DeleteIcon from 'Components/DeleteIcon'

const styles= {
  row:{
    backgroundColor:props=>props.studentError?'rgb(255, 224, 224)':props.studentSuccess?'rgb(223, 239, 216)':'white',
  }
}

const ConversionRow = ({
  //STATE
    fromValue,handleFromUnitSelect,handleFromValueChange,
    toType,toUnit,handleToUnitSelect,
    studentResponse,handleStudentResponseChange,
    studentError,setStudentError,
  //HANDLERS
    calculateResult,handleDelete,
  //OTHER
    classes,...props
})=> {
  let result = calculateResult();
  return (
    <TableRow className={classes.row} data-cy='row'>
      <TableCell>
        <TextField
          className='fromInput'
          value={fromValue}
          onChange={(e)=>handleFromValueChange(e.target.value)}
          type="number"
          InputProps={{disableUnderline:true,"data-cy":'fromValue'}}
          placeholder='value'
          InputLabelProps={{
            shrink: true,
          }}
        />
      </TableCell>
      <TableCell>
        <UnitPicker inputProps={{"data-cy":'fromUnit'}} onChange={handleFromUnitSelect}/>
      </TableCell>
      <TableCell>
        <UnitPicker inputProps={{"data-cy":'toUnit'}} units={toType} value={toUnit} onChange={handleToUnitSelect}/>
      </TableCell>
      <TableCell>
        <TextField
          className='toInput'
          value={studentResponse}
          onChange={(e)=>handleStudentResponseChange(e.target.value)}
          type="number"
          InputProps={{disableUnderline:true,"data-cy":'guess'}}
          placeholder='value'
          InputLabelProps={{
            shrink: true,
          }}
        />
      </TableCell>
      <TableCell>
        <div data-cy='result'>
          {result}
        </div>
      </TableCell>
      <TableCell>
        <DeleteIcon onClick={handleDelete}/>
      </TableCell>
    </TableRow>
  )
}

export default compose(
  withState('fromValue','setFromValue',""),
  withState('fromUnit','setFromUnit',null),
  withState('fromType','setFromType',null),
  withState('toValue','setToValue',null),
  withState('toUnit','setToUnit',null),
  withState('toType','setToType',null),
  withState('studentResponse','setStudentResponse',""),
  withState('studentError','setStudentError',false),
  withState('studentSuccess','setStudentSuccess',false),
  withHandlers({
    //Compare conversion "TO" value to Student Response value and set state
    calculateError:props=>(toValue)=>{
      let shouldShowError = false;
      let shouldShowSuccess = false;
      if(toValue&&props.studentResponse){
        let roundedToValue = math.round(toValue,1)
        let roundedResp = math.round(props.studentResponse,1)
        if(roundedToValue!==roundedResp){
          shouldShowError=true
        }else{
          shouldShowSuccess=true
        }
      }
      if(props.studentError!==shouldShowError){
        props.setStudentError(shouldShowError)
      }
      if(props.studentSuccess!==shouldShowSuccess){
        props.setStudentSuccess(shouldShowSuccess)
      }
    }
  }),
  withHandlers({
    //Calculate Conversion "TO" value
    calculateResult:props=>(overrides)=>{
      let {fromValue,fromUnit,toUnit} = props
      if(overrides){
        fromValue=overrides.fromValue||fromValue
        fromUnit=overrides.fromUnit||fromUnit
        toUnit=overrides.toUnit||toUnit
      }
      if(fromValue&&fromUnit&&toUnit){
        let from = math.unit(fromValue,fromUnit.toLowerCase())
        let toValue = from.toNumber(toUnit.toLowerCase())
        props.calculateError(toValue)
        return toValue&&math.round(toValue,1)
      }else{
        return false
      }
    }
  }),
  withHandlers({
    handleFromUnitSelect:props=>(selectedUnit)=>{
      if(!selectedUnit){//If the selected value is null e.g. the input is cleared
        props.setFromType(null)
        props.setFromUnit(null)
        return;
      }
      props.setFromType(selectedUnit.type)
      props.setFromUnit(selectedUnit.value)
      if(selectedUnit.type!==props.toType){//From and to Unit types must match. Cannot convert temp to volume
        props.setToType(selectedUnit.type)
        props.setToUnit(null)
      }
    },
    handleToUnitSelect:props=>(selectedUnit)=>{
      if(!selectedUnit){//If the selected value is null e.g. the input is cleared
        props.setToType(null)
        props.setToUnit(null)
        return
      }
      props.setToType(selectedUnit.type)
      props.setToUnit(selectedUnit.value)
    },
    handleFromValueChange:props=>(value)=>{
      props.setFromValue(value)
    },
    handleStudentResponseChange:props=>(value)=>{
      props.setStudentResponse(value)
    },
    handleDelete:props=>()=>{
      props.onDelete&&props.onDelete()
    }
  }),
  withStyles(styles)
)(ConversionRow)
