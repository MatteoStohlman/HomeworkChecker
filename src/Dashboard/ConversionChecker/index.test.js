import React from 'react';
import { mount , shallow} from 'enzyme';
import renderer from 'react-test-renderer';
import ConversionChecker from './index';

test('basic ConversionChecker renders and snapshot', () => {
  shallow(<ConversionChecker/>)
  const picker = renderer.create(
    <ConversionChecker/>
  );
  expect(picker.toJSON()).toMatchSnapshot();
});

test('add row functionality works',() => {
  const checker = mount(<ConversionChecker/>)
  let firstRowCount = checker.find("ConversionRow").length
  checker.find('#addRowButton').first().simulate('click')
  let secondRowCount = checker.find("ConversionRow").length
  expect(firstRowCount+1===secondRowCount).toBe(true)
})

test('remove row functionality works',() => {
  const checker = mount(<ConversionChecker/>)
  let firstRowCount = checker.find("ConversionRow").length
  checker.find('DeleteIcon').first().simulate('click')
  let secondRowCount = checker.find("ConversionRow").length
  expect(firstRowCount-1===secondRowCount).toBe(true)
})
