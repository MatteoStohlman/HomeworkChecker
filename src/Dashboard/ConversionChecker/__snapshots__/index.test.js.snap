// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic ConversionChecker renders and snapshot 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded ConvertTable-root-1"
>
  <table
    className="MuiTable-root ConvertTable-table-2"
  >
    <thead
      className="MuiTableHead-root"
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head"
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          From Value
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          From Unit
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          To Unit
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          Student Response
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          Result
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        />
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root"
    >
      <tr
        className="MuiTableRow-root ConversionRow-row-50 ConversionRow-row-51"
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          <div
            className="MuiFormControl-root MuiTextField-root"
          >
            <div
              className="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                className="MuiInputBase-input MuiInput-input"
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="value"
                required={false}
                type="number"
                value=""
              />
            </div>
          </div>
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          <div
            className=" css-2b097c-container"
            onKeyDown={[Function]}
          >
            <div
              className="MuiFormControl-root MuiTextField-root"
            >
              <div
                className="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
                onClick={[Function]}
              >
                <div
                  aria-invalid={false}
                  className="MuiInputBase-input MuiInput-input UnitPicker-input-86"
                  disabled={false}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onMouseDown={[Function]}
                  onTouchEnd={[Function]}
                  required={false}
                  type="text"
                >
                  <div
                    className="UnitPicker-valueContainer-87"
                  >
                    <p
                      className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary UnitPicker-placeholder-90"
                    >
                      pick units
                    </p>
                    <div
                      className="css-1g6gooi"
                    >
                      <div
                        className=""
                        style={
                          Object {
                            "display": "inline-block",
                          }
                        }
                      >
                        <input
                          aria-autocomplete="list"
                          autoCapitalize="none"
                          autoComplete="off"
                          autoCorrect="off"
                          disabled={false}
                          id="react-select-2-input"
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          spellCheck="false"
                          style={
                            Object {
                              "background": 0,
                              "border": 0,
                              "boxSizing": "content-box",
                              "color": "inherit",
                              "fontSize": "inherit",
                              "label": "input",
                              "opacity": 1,
                              "outline": 0,
                              "padding": 0,
                              "width": "1px",
                            }
                          }
                          tabIndex="0"
                          type="text"
                          value=""
                        />
                        <div
                          style={
                            Object {
                              "height": 0,
                              "left": 0,
                              "overflow": "scroll",
                              "position": "absolute",
                              "top": 0,
                              "visibility": "hidden",
                              "whiteSpace": "pre",
                            }
                          }
                        >
                          
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          <div
            className=" css-2b097c-container"
            onKeyDown={[Function]}
          >
            <div
              className="MuiFormControl-root MuiTextField-root"
            >
              <div
                className="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
                onClick={[Function]}
              >
                <div
                  aria-invalid={false}
                  className="MuiInputBase-input MuiInput-input UnitPicker-input-86"
                  disabled={false}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onMouseDown={[Function]}
                  onTouchEnd={[Function]}
                  required={false}
                  type="text"
                >
                  <div
                    className="UnitPicker-valueContainer-87"
                  >
                    <p
                      className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary UnitPicker-placeholder-90"
                    >
                      pick units
                    </p>
                    <div
                      className="css-1g6gooi"
                    >
                      <div
                        className=""
                        style={
                          Object {
                            "display": "inline-block",
                          }
                        }
                      >
                        <input
                          aria-autocomplete="list"
                          autoCapitalize="none"
                          autoComplete="off"
                          autoCorrect="off"
                          disabled={false}
                          id="react-select-3-input"
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          spellCheck="false"
                          style={
                            Object {
                              "background": 0,
                              "border": 0,
                              "boxSizing": "content-box",
                              "color": "inherit",
                              "fontSize": "inherit",
                              "label": "input",
                              "opacity": 1,
                              "outline": 0,
                              "padding": 0,
                              "width": "1px",
                            }
                          }
                          tabIndex="0"
                          type="text"
                          value=""
                        />
                        <div
                          style={
                            Object {
                              "height": 0,
                              "left": 0,
                              "overflow": "scroll",
                              "position": "absolute",
                              "top": 0,
                              "visibility": "hidden",
                              "whiteSpace": "pre",
                            }
                          }
                        >
                          
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          <div
            className="MuiFormControl-root MuiTextField-root"
          >
            <div
              className="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                className="MuiInputBase-input MuiInput-input"
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="value"
                required={false}
                type="number"
                value=""
              />
            </div>
          </div>
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          <span
            aria-hidden={true}
            className="material-icons MuiIcon-root DeleteIcon-root-123"
            onClick={[Function]}
          >
            delete
          </span>
        </td>
      </tr>
    </tbody>
  </table>
  <button
    className="MuiButtonBase-root MuiFab-root MuiFab-primary ConvertTable-addRowButton-3"
    disabled={false}
    id="addRowButton"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiFab-label"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
    </span>
  </button>
</div>
`;
